src/newscale/stages.py

```diff
@@ -1,365 +1,497 @@
-        is set to 0.
-        NOTE: is there really no way to do this idempotently? Or for that matter,
-        to query the current status?
-        """
-        cmd = "<07>\r"
-        cmd_bytes = bytes(cmd, 'utf-8')
-        self.sock.sendall(cmd_bytes)
-        resp = self.sock.recv(1024)
-
-    # 08
-    def moveToTarget(self, targetValue):
-        """
-        This command sets a target position and moves the motor to that target
-        position at the speed defined by command <40>.
-        ONLY VALID IN CLOSED-LOOP MODE.
-        """
-        targetValue = int(targetValue)
-        cmd = "<08 {0:08x}>\r".format(targetValue)
-        cmd_bytes = bytes(cmd, 'utf-8')
-        self.sock.sendall(cmd_bytes)
-        resp = self.sock.recv(1024)
-
-    # 09
-    def setOpenLoopSpeed(self, speed_255):
-        """
-        This command sets the open-loop speed of the motor, as a range from 0-255,
-        with 255 representing 100% speed. This value is not saved to internal EEPROM.
-        """
-        speed_255 = int(speed_255)
-        cmd = "<09 {0:02x}>\r".format(speed_255)
-        cmd_bytes = bytes(cmd, 'utf-8')
-        self.sock.sendall(cmd_bytes)
-        resp = self.sock.recv(1024)
-
-    # 10
-    def viewClosedLoopStatus(self):
-        """
-        This command is used to view the motor status and position.
-        """
-        cmd = "<10>\r"
-        cmd_bytes = bytes(cmd, 'utf-8')
-        self.sock.sendall(cmd_bytes)
-        resp = self.sock.recv(1024).decode('utf-8').strip('<>\r')
-        SSSSSS = int(resp.split()[1], 16)
-        PPPPPPPP = int(resp.split()[2], 16)
-        EEEEEEEE = int(resp.split()[3], 16)
-        return SSSSSS, PPPPPPPP, EEEEEEEE
-        
-    # 19
-    def readMotorFlags(self):
-        """
-        This command reports internal flags used by the controller to monitor
-        motor conditions.
-        """
-        cmd = "<19>\r"
-        cmd_bytes = bytes(cmd, 'utf-8')
-        self.sock.sendall(cmd_bytes)
-        resp = self.sock.recv(1024).decode('utf-8').strip('<>\r')
-        # TODO parse response
+        return self._send(self._get_cmd_str(Cmd.HALT))
+
+    # <04>
+    def run(self, direction: Direction, seconds: float = None):
+        """Enable motor for a specified time or without limit if unspecified.
+
+        :param direction: forward or reverse
+        :param seconds: time to run the motor in seconds (with increments of
+            tenths of seconds).
+        """
+        assert direction is not Direction.NEITHER, \
+            "Direction must be forward or reverse."
+        if seconds is None:
+            return self._send(self._get_cmd_str(Cmd.RUN, direction.value))
+        duration = round(seconds*10)  # value is encoded in tenths of seconds.
+        assert duration < 256, "Time input value exceeds maximum value."
+        return self._send(self._get_cmd_str(Cmd.RUN, direction.value,
+                                            f"{duration:04x}"))
+
+    # <05>
```

Comment:
any reason why this isn't implemented?

Reply:
The short answer: I was too lazy to understand/test what was going on to wrap it into a function. It looks like we also have to issue a <52> also to figure out what the units of the arguments are? Do we really need it for a first pass?

Reply:
i can accept that! i just checked and we are not using it currently in parallax/mis-guide, so no we don't need it for a first pass.